### PR TITLE
Issue #2807: FinalLocalVariable doesn't report variable when conditio…

### PR DESCRIPTION
**Pull request by Bhavik3**
…n separates two assignments

### CODE REVIEW

**Comment by romani** [comment]
1) please send PR to our fork of Orekit project.

2) please fix coverage

3) please split your commit into two, first is changes to missed finals, second is functional changes in Check. This will help us to clearly review Check changes and let you apply codereview changes without problems.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FinalLocalVariableCheck.java [hunk 1]
```diff
@@ -223,10 +228,20 @@ public void leaveToken(DetailAST ast) {
...
+                    prevScopeUninitializedVariableStack.peek();
                 if (ast.getParent().getType() != TokenTypes.CASE_GROUP
-                    || findLastToken(ast.getParent().getParent(), TokenTypes.CASE_GROUP,
-                        TokenTypes.SLIST) == ast.getParent()) {
+                    || findLastChildWhichContainsSpecifiedToken(ast.getParent().getParent(),
+                            TokenTypes.CASE_GROUP, TokenTypes.SLIST) == ast.getParent()) {
                     scope = scopeStack.pop().scope;
+                    prevScopeUninitializedVariableStack.pop();
+                }
+
```
Comment (mkordas) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FinalLocalVariableCheck.java]: @Bhavik3 I don't think we should have empty space here, as It should be as close to `case` as possible

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FinalLocalVariableCheck.java [hunk 2]
```diff
@@ -240,6 +255,64 @@ public void leaveToken(DetailAST ast) {
...
+                final DetailAST storedVariable = scopeData.scope.get(variable.getText());
+                if (storedVariable != null && isSameVariables(storedVariable, variable)) {
+                    scopeData.uninitializedVariables.push(variable);
+                }
+            }
+        }
+    }
+
+    /**
+     * Check weather token is LITERAL_TRY, LITERAL_CATCH or LITERAL_FINALLY.
```
Comment (mkordas) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FinalLocalVariableCheck.java]: @Bhavik3 typo: `weather` :)

**Comment by Bhavik3** [comment]
@romani 

1) Done

2) Done

3) Done

**Comment by romani** [comment]
@Bhavik3 , thanks a lot for your help.

changes are major, we need diff report of violation before and after fix (by ahsm tool).
While I am reading the code , please generate reports.

FYI: it will help you a lot to swap order of commits in your branch to have changes to Check be the last commit to ease fixes in it, commit with "final" to our code is unlikely to change.

**Comment by Bhavik3** [comment]
@romani 

Reports

Material Design : http://bhavik3.github.io/result_Material_design/checkstyle_merged.html

Orekit : http://bhavik3.github.io/result_Orekit/checkstyle_merged.html

Android App launcher : http://bhavik3.github.io/result_android_app_launcher/checkstyle_merged.html

PMD : http://bhavik3.github.io/result_pmd/checkstyle_merged.html

Java Design Pattern : http://bhavik3.github.io/result_Java_design_pattern/checkstyle_merged.html

Reports are fine.

**Comment by romani** [comment]
reports are fine , thanks a lot !!

I will finish review soon.

**Comment by mkordas** [comment]
@romani @Bhavik3 I'm fine with changes, thanks!

**Comment by romani** [comment]
there is a bit more cases that you touched..... but lets resolve them now, while I see them.

1) 
> prevScopeUninitializedVariableStack

please remove "Stack" and make name in plural form

2)
> `Deque<PrevScopeUnitializedVariableData>`

Long of class is too long and name duplication in variable looks suspicious. One more problem is class with one filed. I think it is much easier to remove PrevScopeUnitializedVariableData class and just use current variable name to explain the meaning. collection will be just `Deque<Deque<DetailAST>>`.

3)
>`*     &lt;property name="tokens" value="VARIABLE_DEF"/&gt;`

2 cases there javadoc show incorrect configuration - all required tokens have to be provided in case of overriding default.

4) 
>```
                if (!isTryCatchOrFinallyToken(parent)
                        && parent.getType() != TokenTypes.LITERAL_ELSE) {
```

Why that parts are not combined in one method and they do the same - checking parent .
New method should explain a reason of IF condition by its name.

5)
> private void storePrevScopeUninitializedVariableData() {

You just copy all variables to separate collection, that collection is never changed after this , I do see a good reason of such cache. Please change or explain.

6)
> final PrevScopeUnitializedVariableData prevScopeUnitializedVariableData =
                    prevScopeUninitializedVariableStack.peek();

It is no clear why you validating only previous scope, I think you have to consider to use all scopes of the stack.

7)
> private void updateUninitializedVariable(PrevScopeUnitializedVariableData
            prevScopeUnitializedVariableData) {

put "s" to the end of method , as not only one variable is under consideration.

8)
>
```
+            "288:13: " + getCheckMessage(MSG_KEY, "a"),
+            "300:13: " + getCheckMessage(MSG_KEY, "a"),
+            "344:13: " + getCheckMessage(MSG_KEY, "x"),
```

please rename variables `a` and `x` to shouldBeFinal it will be much clear test case code and UTs.


9)
> InputFinalLocalVariable.java

Please keep all this simple test cases and add new test cases (as extension to these) with few scopes between declaration and definition (inner IF, inner SWITCH, inner TRY-CATCH, ....)

10)
> // No Violation

please update all your test cases to have comments like this, to make it clear where violation is expected by quick look to your code.

11)
after fixes, please generate diff report for the same projects and compare with existing, please verify if diff-reports are different.

**Comment by Bhavik3** [comment]
@romani 

3) 9) I didn't understand clearly.

**Comment by romani** [comment]
3)

skip it , my fault.

9)

related to "6)". Please add cases with extra scope between declaration and usage. Example:
```
    public void test() {
        final boolean b = true;
        int a;        //Violation
        if (field) { //<---  additional scope
          if (b) {
            a = 1;
          }
          else {
            a = 2;
          }
     }
    }
```

please do the same for all other cases that you added.

**Comment by romani** [comment]
@Bhavik3 , release will in 10 days(31th January), it would be good to finish this PR.

**Comment by Bhavik3** [comment]
@romani 

1) Done
2) Done
3) skipped
4) Done
5) yes, I am not changing it afterward. Just storing it for future use.
Using cache. how can I use cache ?
I do not understand
6) You are right. Should consider all previous stack.
 Done
7) Done
8) Done
9) same as 6) .. Done
10) Done

**Comment by romani** [comment]
5)
>
```                     scope = scopeStack.pop().scope;
                    prevScopeUninitializedVariables.pop();
```

and
```
for (DetailAST variable : scopeData.uninitializedVariables) {
            prevScopeUnitializedVariableData.push(variable);
        }
```

By means of cache I meant `this.prevScopeUninitializedVariables` collection. You store in it complete copy of `this.scopeStack.peek().uninitializedVariables`. 

Can you remove prevScopeUninitializedVariables collection from code ? if not please show me a place where it is not enough to have this.scopeStack.peek().uninitializedVariables.

As soon as we resolve this point, please generate diff reports again to make sure there is no Exceptions and false-positives. All other code is good for merge.

**Comment by Bhavik3** [comment]
@romani 

consider this example:
```
int x;
if (condition) {
    x++; // when this will be checked, x will be removed from prev scopeStack.uninitializedVaribales, but not from prevScopeUninitializedVariables
}
```

When I call updateUninitializedVariables() , Need information about old uninitialized variables, not new one. that is why we need to store old one in collection.

**Comment by romani** [comment]
scopeStack is a "Stack" it has all previous scopes information. peek of this collection will be new scope - yes, but you can access to all previous scopes data.

Please point more precisely to a line of code that demand you have extra stack for this.

**Comment by Bhavik3** [comment]
@romani 

```
private void removeVariable(DetailAST ast) {
         final Iterator<ScopeData> iterator = scopeStack.descendingIterator();
         while (iterator.hasNext()) {
             final ScopeData scopeData = iterator.next();
             final Map<String, DetailAST> scope = scopeData.scope;
             final DetailAST storedVariable = scope.get(ast.getText());
             if (storedVariable != null && isSameVariables(storedVariable, ast)) {
                 if (shouldRemoveVariable(scopeData, ast)) {
                     scope.remove(ast.getText());
                 }
                 break;
             }
         }
     }
```
Here, code will iterate through scopeStack and will call ```shouldRemoveVariable(scopeData, ast)``` .

Here is ```shouldRemoveVariable(scopeData, ast)``` method.
private static boolean shouldRemoveVariable(ScopeData scopeData, DetailAST ast) {
       .

                 scopeData.uninitializedVariables.remove(variable);
                 break;
             }
         }
         return shouldRemove;
     }

**Comment by romani** [comment]
Please prepare html reports for final verification before merge

**Comment by Bhavik3** [comment]
@romani

Reports

Material Design : http://bhavik3.github.io/result_Material_design/checkstyle_merged.html

Orekit : http://bhavik3.github.io/result_Orekit/checkstyle_merged.html

Android App launcher : http://bhavik3.github.io/result_android_app_launcher/checkstyle_merged.html

PMD : http://bhavik3.github.io/result_pmd/checkstyle_merged.html

Java Design Pattern : http://bhavik3.github.io/result_Java_design_pattern/checkstyle_merged.html

Reports are fine.

**Comment by romani** [comment]
merged as FF.

thanks a lot for your hard work !!!

**Pull request closed without merging**